File: trunk/proj/Android/src/com/jasonzqshen/familyaccounting/core/transaction/DocumentIdentity.java
package com.jasonzqshen.familyaccounting.core.transaction;

import com.jasonzqshen.familyaccounting.core.exception.DocumentIdentityFormatException;
import com.jasonzqshen.familyaccounting.core.exception.FiscalMonthRangeException;
import com.jasonzqshen.familyaccounting.core.exception.FiscalYearRangeException;
import com.jasonzqshen.familyaccounting.core.exception.IdentityInvalidChar;
import com.jasonzqshen.familyaccounting.core.exception.IdentityNoData;
import com.jasonzqshen.familyaccounting.core.exception.IdentityTooLong;
import com.jasonzqshen.familyaccounting.core.exception.MonthIdentityFormatException;

public class DocumentIdentity {
	public final DocumentNumber _docNumber;
	public final MonthIdentity _monthIdentity;

	public DocumentIdentity(DocumentNumber docNum, int fiscalYear,
			int fiscalMonth) throws FiscalYearRangeException,
			FiscalMonthRangeException {
		_docNumber = docNum;
		_monthIdentity = new MonthIdentity(fiscalYear, fiscalMonth);
	}

	public DocumentIdentity(DocumentNumber docNum, MonthIdentity monthIdentity) {
		_docNumber = docNum;
		_monthIdentity = monthIdentity;
	}

	@Override
	public boolean equals(Object obj) {
		if (!(obj instanceof DocumentIdentity)) {
			return false;
		}

		DocumentIdentity id = (DocumentIdentity) obj;
		boolean ret = id._docNumber.equals(_docNumber)
				&& id._monthIdentity.equals(_monthIdentity);
		return ret;
	}

	@Override
	public int hashCode() {
		String str = this.toString();
		int count = 0;
		for (int i = 0; i < str.length(); ++i) {
			count += str.charAt(i);
		}

		return count;
	}

	@Override
	public String toString() {
		return String.format("%s_%s", _docNumber, _monthIdentity);
	}

	/**
	 * parse the string to the document identity
	 * 
	 * @param string
	 * @return
	 * @throws DocumentIdentityFormatException
	 */
	public static DocumentIdentity parse(String string)
			throws DocumentIdentityFormatException {
		String docNumStr = string.substring(0, 10);
		String monthIdStr = string.substring(11, 18);

		try {
			DocumentNumber id = new DocumentNumber(docNumStr.toCharArray());
			MonthIdentity monthId = MonthIdentity.parse(monthIdStr);

			return new DocumentIdentity(id, monthId);
		} catch (IdentityTooLong e) {
			throw new DocumentIdentityFormatException();
		} catch (IdentityNoData e) {
			throw new DocumentIdentityFormatException();
		} catch (IdentityInvalidChar e) {
			throw new DocumentIdentityFormatException();
		} catch (NumberFormatException e) {
			throw new DocumentIdentityFormatException();
		} catch (MonthIdentityFormatException e) {
			throw new DocumentIdentityFormatException();
		}
	}
}

File: trunk/proj/WinStore_FamilyAccounting/WinStore_FamilyAccountingCore/Transaction/DocumentIdentity.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using WinStore_FamilyAccountingCore.Exceptions;
using WinStore_FamilyAccountingCore.Exceptions.FormatExceptions;

namespace WinStore_FamilyAccountingCore.Transaction
{
    /// <summary>
    /// Document identity
    /// </summary>
    public class DocumentIdentity
    {
        public readonly DocumentNumber _docNumber;
        public readonly MonthIdentity _monthIdentity;

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="docNum"></param>
        /// <param name="fiscalYear"></param>
        /// <param name="fiscalMonth"></param>
        /// <exception cref="FiscalYearRangeException"></exception>
        /// <exception cref="FiscalMonthRangeException"></exception>
        public DocumentIdentity(DocumentNumber docNum, int fiscalYear,
                int fiscalMonth)
        {
            _docNumber = docNum;
            _monthIdentity = new MonthIdentity(fiscalYear, fiscalMonth);
        }

        /// <summary>
        /// constructor
        /// </summary>
        /// <param name="docNum"></param>
        /// <param name="monthIdentity"></param>
        public DocumentIdentity(DocumentNumber docNum, MonthIdentity monthIdentity)
        {
            _docNumber = docNum;
            _monthIdentity = monthIdentity;
        }

        /// <summary>
        /// 
        /// </summary>
        /// <param name="obj"></param>
        /// <returns></returns>
        public override bool Equals(Object obj)
        {
            DocumentIdentity id = obj as DocumentIdentity;
            if (id == null)
            {
                return false;
            }
            return id._docNumber.Equals(_docNumber)
                    && id._monthIdentity.Equals(_monthIdentity);
        }
        /// <summary>
        /// Get hash code
        /// </summary>
        /// <returns></returns>
        public override int GetHashCode()
        {
            String str = this.ToString();
            int count = 0;
            for (int i = 0; i < str.Length; ++i)
            {
                count += str[i];
            }

            return count;
        }

        /// <summary>
        /// to string
        /// </summary>
        /// <returns></returns>
        public override String ToString()
        {
            return String.Format("{0}_{1}", _docNumber, _monthIdentity);
        }

        /// <summary>
        /// Parse string to Document Identity
        /// </summary>
        /// <param name="docIdStr"></param>
        /// <returns></returns>
        /// <exception cref="DocumentIdentityFormatException">Document Identity Exception</exception>
        public static DocumentIdentity Parse(String docIdStr)
        {
            try
            {
                String docNumStr = docIdStr.Substring(0, 10);
                String monthIdStr = docIdStr.Substring(11, 7);

                DocumentNumber id = new DocumentNumber(docNumStr.ToCharArray());
                MonthIdentity monthId = MonthIdentity.Parse(monthIdStr);

                return new DocumentIdentity(id, monthId);
            }
            catch (ArgumentOutOfRangeException)
            {
                throw new DocumentIdentityFormatException(docIdStr);
            }
            catch (IdentityTooLong)
            {
                throw new DocumentIdentityFormatException(docIdStr);
            }
            catch (IdentityNoData)
            {
                throw new DocumentIdentityFormatException(docIdStr);
            }
            catch (IdentityInvalidChar)
            {
                throw new DocumentIdentityFormatException(docIdStr);
            }
            catch (MonthIdentityFormatException)
            {
                throw new DocumentIdentityFormatException(docIdStr);
            }
        }
    }
}
